Add bare-metal examples

e-rk opened the pull request:
Bare-metal examples for the radio driver have been added.

TODO:

- [x] Example readme

hubertmis commented on src/raal/simulator/nrf_raal_simulator.c:
```diff
@@ -46,8 +46,6 @@
 #include "nrf.h"
 #include "nrf_raal_config.h"
 
-#include <cmsis/cmsis_gcc.h>
```
It seems that header file defining NVIC_* function is missing now. Isn't it?

LuDuda replied:
nrf.h includes nrf52.h which includes proper cmsis file which is cmsis_core4.h

hubertmis replied:
Is it possible to include it directly?

LuDuda replied:
Isn't that idea of nrf.h to include all chip related functionality? But okay i will include it.

hubertmis replied:
I'm OK to keep this include indirectly through `nrf.h` in this case.

e-rk committed:
Rename directories holding SL library files

The paths have been changed to lower-case for ease of handling
in the build system.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

e-rk committed:
Use path relative to SL include root to include platform headers

Previously the platform files inconsistently included platform headers
using either path relative to include root or relative to source file's
directory. This inconsistency could cause build errors under certain
circumstances.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

e-rk committed:
Add bare-metal examples

This change adds examples presenting the use of nRF IEEE 802.15.4 radio driver
in bare metal application.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

hubertmis commented on CMakeLists.txt:
```diff
@@ -0,0 +1,105 @@
+cmake_minimum_required(VERSION 3.13.1)
```
Is it safe to add CMakeLists.txt to top directory of repository that is going to be a part of NCS? I'm not sure if we can expect any potential conflicts in build system IDEs, etc.

e-rk replied:
I don't see any reason for it to cause any problems. Zephyr modules do not magically pull any CMakeLists file unless explicitly told to do so. This file can be pulled in a controlled way in an equivalent fashion as with OpenThread
https://github.com/zephyrproject-rtos/openthread/blob/1d668284a0b897bd6d539e7c3ed8009f0028099d/zephyr/CMakeLists.txt#L229

hubertmis replied:
Most probably my doubt is caused by the CMakeLists file used in Zephyr that was present in hal_nordic repository in top 802.15.4 directory. But now as this file is going to be moved to Zephyr upstream tree, I don't see conflicts anymore. Thanks for explaining.

hubertmis commented on external/cmsis/arm_common_tables.h:
```diff
@@ -0,0 +1,136 @@
+/* ----------------------------------------------------------------------
```
Is there any reason to copy-pste CMSIS in here instead of using submodules like for nrfx and nrfxlib?

e-rk replied:
Done. I have used 4.5.0 tag for the submodule.

hubertmis reviewed (changes requested):
I like that examples are finally coming to this repository.
I'm curious if build system added for these examples can be conflicting in Zephyr / NCS build systems. Also I'm thinking if we could remove a copy of CMSIS from this repository

jciupis commented on examples/README.md:
```diff
@@ -0,0 +1,134 @@
+target_xyz(...)
+
+#
+# Include a CMake file which sets NRF_802154_DRIVER_PLATFORM_BAREMETAL
```
If that's an option, I'd be in favour of renaming this CMake variable to `NRF_802154_DRIVER_SOURCES_PLATFORM_BAREMETAL` so that:
* the name clearly implies a list of source files, not e.g. a boolean that indicates whether platform type is baremetal
* the name matches its SL counterpart

e-rk replied:
Variable was renamed as suggested.

jciupis commented on examples/README.md:
```diff
@@ -0,0 +1,134 @@
+#
+# nRF52840 with hard-float ABI is used as an example.
+#
+set(TARGET_SOC "nrf52840")
```
I'd start the example CMake integration with setting these variables, because the list of steps mentioned earlier starts with that as well:

> The CMake description that pulls the radio driver must set the following parameters prior to the directory inclusion:
> * `TARGET_SOC` variable - specifies the SoC that the radio driver is meant to be built for.
> * `TARGET_ABI` variable - specifies the float ABI that the radio driver is meant to use.

This way it's crystal clear that the CMake code is an example implementation of the aforementioned steps.

e-rk replied:
The intention behind this approach for examples is to make it easy for the user to specify a platform.
For example the user can do:
```
cmake <dir> -DTARGET_SOC=nrf52811
```
and not have to specify ABI separately, which might result in non-functional firmware.

jciupis commented on examples/nrf_802154_delayed_echo_client/src/nrf_802154_delayed_echo_client.c:
```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2018 - 2020, Nordic Semiconductor ASA
+ * All rights reserved.
+ *
```
Missing SPDX tag. Applies to source files of the remaining examples as well

e-rk replied:
Done.

jciupis commented on examples/nrf_802154_delayed_echo_client/src/nrf_802154_delayed_echo_client.c:
```diff
@@ -0,0 +1,234 @@
+        __WFE();
+    }
+
+    return 0;
```
Unreachable code. Applies to source files of the remaining examples as well

e-rk replied:
Return statement was removed.

jciupis reviewed (commented):
Fantastic job :+1: some minor comments below

pkwiek commented on CMakeLists.txt:
```diff
@@ -0,0 +1,107 @@
+add_library(nrf_802154_radio_driver STATIC "")
+target_sources(nrf_802154_radio_driver
+  PRIVATE
+    ${NRF_802154_DRIVER_SOURCES_COMMON}
```
`${NRF_802154_DRIVER_SOURCES_COMMON}` file list is included in `${NRF_802154_DRIVER_SOURCES_NRF52}`.

e-rk replied:
Removed this line.

e-rk commented:
Rebased and smoke tested.

pkwiek commented on examples/nrf_802154_delayed_echo_client/src/nrf_802154_delayed_echo_client.c:
```diff
@@ -0,0 +1,235 @@
+/*
```
It would be good to run uncrustify with driver's `.uncrustify.cfg` on `nrf_802154_delayed_echo_client.c`, `nrf_802154_delayed_echo_server.c`, `nrf_802154_echo_client.c`, `nrf_802154_echo_server.c` and `mpsl_utils.c` files.

e-rk replied:
Done.

pkwiek reviewed (approved):
Great!
